scheduler/src/tbf/ratelimit.clj

```diff
@@ -0,0 +1,101 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns tbf.ratelimit
+  (:require
+    [tbf.core :as tbf-ns])
+  (:import (com.google.common.cache CacheBuilder Cache)
+           (java.util.concurrent TimeUnit)))
+
+(defprotocol RateLimiter
+  (earn-tokens [this key]
+    "Account for any earned tokens for the requested key.")
+  (in-debt [this key]
```

Comment:
It seems strange that we need to explicitly call `earn-tokens` before `in-debt` returns a usable value.

Reply:
I put in ? and ! in this and the other fields to indicate the mutation and functionality. It was done this way to reduce at least one form of API error. We could make it be (in-debt!? [this key]]) and explicitly mutate then return the predicate?

Reply:
I'm okay with that. I do not see why we would want to use `debt?`